drafts/str4d-overwinter-sighash/draft1.rst

```diff
@@ -0,0 +1,312 @@
+  [#offline-wallets]_
+
+Deploying the aforementioned fixes in the original script system is not a simple task.
+
+
+Specification
+=============
+
+A new transaction digest algorithm is defined, but only applicable from the Overwinter upgrade block height
+[#ZIP0000]_::
+  BLAKE2b-256 of the serialization of:
+    1. nVersion of the transaction (4-byte little endian)
+    2. nVersionBranch of the transaction (4-byte little endian)
+    3. hashPrevouts (32-byte hash)
+    4. hashSequence (32-byte hash)
+    5. hashOutputs (32-byte hash)
+    6. hashJoinSplits (32-byte hash)
+    7. nLockTime of the transaction (8-byte little endian)
+    8. nExpiryTime of the transaction (8-byte little endian)
+    9. sighash type of the signature (4-byte little endian)
+   10. If we are serializing an input (ie. this is not a JoinSplit signature hash):
+       a. outpoint (32-byte hash + 4-byte little endian) 
+       b. scriptCode of the input (serialized as scripts inside CTxOuts) [TODO]
+       c. value of the output spent by this input (8-byte little endian)
+       d. nSequence of the input (4-byte little endian)
+
+The BLAKE2b-256 personalization field is set to::
+
+  "ZcashSigHash" || BLOCK_BRANCH_ID
+
+``BLOCK_BRANCH_ID`` is the ``BRANCH_ID`` for the epoch of the transaction's parent block. [#ZIP-activation-mechanism]_
+Domain separation of the signature hash across parallel branches provides replay protection: transactions
+targeted for one branch will have invalid signatures on other branches. This has two effects on user behavior:
+
+* Transaction creators MUST specify the epoch they want their transaction to be mined in. Across a network
+  upgrade, this means that if a transaction is not mined before the activation height, it will never be mined.
+
+* Transaction validators MUST NOT use ``nVersionBranch`` as the ``BLOCK_BRANCH_ID``: while these are both a
+  ``BRANCH_ID`` "under the hood", they are **semantically different**.
+
+Semantics of the original sighash types remain unchanged, except the following:
+
+#. The serialization format is changed;
+
+#. All sighash types commit to the amount being spent by the signed input;
+
+#. ``SINGLE`` does not commit to the input index. When ``ANYONECANPAY`` is not set, the semantics are
+   unchanged since ``hashPrevouts`` and ``outpoint`` together implictly commit to the input index. When
+   ``SINGLE`` is used with ``ANYONECANPAY``, omission of the index commitment allows permutation of the
+   input-output pairs, as long as each pair is located at an equivalent index.
+
+Field definitions
+-----------------
+
+The items 1, 7, 9, 10a, 10d have the same meaning as the original algorithm. [#wiki-checksig]_
+
+2: ``nVersionBranch``
+`````````````````````
+The ``BRANCH_ID`` of the epoch in which the transaction format corresponding to ``nVersion`` was introduced.
+[#ZIP-overwinter-tx-format]_
+
+3: ``hashPrevouts``
```

Comment:
So most of this document here is the same as  BIP143 and this should be made clearer at start of ZIP. 
 It would be helpful if we clearly labelled which sections are Zcash only, which are pure BIP143 and which are BIP143+ZcashModification  This will help folk who are familiar with that BIP.

Reply:
Added a note about this being based on BIP 143. Note that none of the document is pure BIP 143 anymore, so while some code reuse is possible, it will likely be of the copy-paste form (e.g. `hashPrevouts` now uses BLAKE2b, not SHA256d, so a function for calculating that can't trivially be reused).